VxWorks/Os/Queue.cpp

```diff
@@ -1,96 +1,105 @@
+#include "Queue.hpp"
+#include <Fw/Types/Assert.hpp>
 
-#include <vxWorks.h>
-#include <msgQLib.h>
+namespace Os {
+namespace VxWorks {
+namespace Queue {
 
-#include <stdio.h>
-#include <string.h>
-#include <logLib.h>
+VxWorksQueue::~VxWorksQueue() {
+    (void)msgQDelete(this->m_handle.m_queue);
+}
 
-namespace Os {
-    
-    Queue::Queue(): m_handle(0) {
-        
+QueueInterface::Status VxWorksQueue::create(const Fw::StringBase& name, FwSizeType depth, FwSizeType messageSize) {
+    this->m_handle.m_queue = msgQCreate(depth, messageSize, MSG_Q_PRIORITY);
+    if (this->m_handle.m_queue == MSG_Q_ID_NULL) {
+        return QueueInterface::Status::UNINITIALIZED;
+    }
+    return QueueInterface::Status::OP_OK;
+}
+
+QueueInterface::Status VxWorksQueue::send(const U8* buffer,
+                                          FwSizeType size,
+                                          FwQueuePriorityType priority,
+                                          QueueInterface::BlockingType blockType) {
+    FW_ASSERT(buffer != nullptr);
+    if (this->m_handle.m_queue == MSG_Q_ID_NULL) {
+        return QueueInterface::Status::UNINITIALIZED;
     }
 
-    Queue::QueueStatus Queue::createInternal(const Fw::StringBase &name, NATIVE_INT_TYPE depth, NATIVE_INT_TYPE msgSize) {
+    PlatformIntType vxPrio = (priority > 0) ? MSG_PRI_URGENT : MSG_PRI_NORMAL;
 
-        this->m_name = "QV_";
-        this->m_name += name;
+    // Doing a c-style cast here because msgQSend requires a char* and this is more
+    // efficient than copying from a const U8 buffer to a non-const char* buffer.
+    STATUS stat = msgQSend(this->m_handle.m_queue, (char*)buffer, size,
+                           (QueueInterface::BlockingType::NONBLOCKING == blockType) ? NO_WAIT : WAIT_FOREVER, vxPrio);
 
-        MSG_Q_ID handle = msgQCreate(depth,msgSize,MSG_Q_PRIORITY);
-        this->m_handle = handle == NULL?0:(POINTER_CAST)handle;
-        if (NULL == handle) {
-            return QUEUE_UNINITIALIZED;
+    if (stat == VXWORKS_ERROR) {
+        switch (errno) {
+            case S_msgQLib_INVALID_MSG_LENGTH:
+                return QueueInterface::Status::SIZE_MISMATCH;
+            case S_objLib_OBJ_UNAVAILABLE:
+            case S_msgQLib_NON_ZERO_TIMEOUT_AT_INT_LEVEL:
+                return QueueInterface::Status::FULL;
+            default:
+                return QueueInterface::Status::UNKNOWN_ERROR;
         }
-        Queue::s_numQueues++;
-        return QUEUE_OK;
-    }
-    
-    Queue::~Queue() {
-        (void)msgQDelete((MSG_Q_ID) this->m_handle);
     }
 
-    Queue::QueueStatus Queue::send(const U8* buffer, NATIVE_INT_TYPE size, NATIVE_INT_TYPE priority, QueueBlocking block) {
+    // Protect critical data m_highMark
+    {
+        Os::ScopeLock lock(const_cast<Mutex&>(this->m_handle.m_data_lock));
+        this->m_handle.m_highMark = FW_MAX(this->m_handle.m_highMark, this->getMessagesAvailable());
```

Comment:
I'm not sure I'm a fan of adding the overhead of the creation of the `Os::ScopeLock` and locking on every message insertion. Could we at least wrap it in a macro so we can turn it off? I wouldn't hold back the PR, but we should talk about it as a group.

Reply:
The high watermark is a feature expected of the Queue Interface. If we want to turn this on or off, it'll require updates to the core F Prime Queue layer.

For now, we'll leave this as is and update it once F Prime has a good solution to turn this high watermark feature on or off.